fabric-api-lookup-api-v1/src/main/java/net/fabricmc/fabric/impl/lookup/ApiLookupMapImpl.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.impl.lookup;
+
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+import net.minecraft.util.Identifier;
+
+import net.fabricmc.fabric.api.lookup.v1.ApiLookupMap;
+
+public final class ApiLookupMapImpl<L> implements ApiLookupMap<L> {
+	private final Map<Identifier, StoredLookup<L>> lookups = new HashMap<>();
+	private final Supplier<L> lookupFactory;
+
+	public ApiLookupMapImpl(Supplier<L> lookupFactory) {
+		this.lookupFactory = lookupFactory;
+	}
+
+	@Override
+	public synchronized L getLookup(Identifier lookupId, Class<?> apiClass, Class<?> contextClass) {
+		StoredLookup<L> storedLookup = lookups.computeIfAbsent(lookupId, id -> new StoredLookup<>(lookupFactory.get(), apiClass, contextClass));
+
+		if (storedLookup.apiClass == apiClass && storedLookup.contextClass == contextClass) {
+			return storedLookup.lookup;
+		}
+
+		final String errorMessage = String.format(
+				"Lookup with id %s is already registered with api class %s and context class %s. It can't be registered with api class %s and context class %s.",
+				lookupId,
+				storedLookup.apiClass.getCanonicalName(),
+				storedLookup.contextClass.getCanonicalName(),
+				apiClass.getCanonicalName(),
+				contextClass.getCanonicalName()
+		);
+
+		throw new IllegalArgumentException(errorMessage);
+	}
+
+	@Override
+	public synchronized Iterator<L> iterator() {
+		return lookups.values().stream().map(storedLookup -> storedLookup.lookup).collect(Collectors.toList()).iterator();
+	}
+
+	private static class StoredLookup<L> {
+		private final L lookup;
+		private final Class<?> apiClass;
+		private final Class<?> contextClass;
+
+		private StoredLookup(L lookup, Class<?> apiClass, Class<?> contextClass) {
```

Comment:
```suggestion
	private static final class StoredLookup<L> {
		final L lookup;
		final Class<?> apiClass;
		final Class<?> contextClass;

		StoredLookup(L lookup, Class<?> apiClass, Class<?> contextClass) {
```
This ain't java 11, no nest host yet

Reply:
No idea what the difference is. I don't think `final` is important for an impl class?

Reply:
1. `final` can make jvm better optimize pojos
2. Removing `private` from inner class members (not class itself, as the class modifiers are effectively package and public) allows the compiler to not generate bridge methods for each place where the private field is accessed or the private constructor is called. fyi the accessor in compiled java looks like `access$$0` something, while the constructors have some useless arguments inserted by compiler to avoid conflicts. 
   - In Java 11, the accessor problem is gone as jvm would recognize nest mates for private access and compiler no longer has to generate bridges. But remember, we are on Java 8

Reply:
I'm pretty sure the JVM already recognizes effectively `final` classes, but this change can't hurt anyway, and sure it's a good idea to remove synthetic methods as well. :)